src/crypto/index.js

```diff
@@ -377,22 +380,26 @@ function _uploadOneTimeKeys(crypto) {
     const oneTimeKeys = crypto._olmDevice.getOneTimeKeys();
     const oneTimeJson = {};
 
+    const promises = [];
+
     for (const keyId in oneTimeKeys.curve25519) {
         if (oneTimeKeys.curve25519.hasOwnProperty(keyId)) {
             const k = {
                 key: oneTimeKeys.curve25519[keyId],
             };
-            crypto._signObject(k);
             oneTimeJson["signed_curve25519:" + keyId] = k;
+            promises.push(crypto._signObject(k));
         }
     }
 
-    return crypto._baseApis.uploadKeysRequest({
-        one_time_keys: oneTimeJson,
-    }, {
-        // for now, we set the device id explicitly, as we may not be using the
-        // same one as used in login.
-        device_id: crypto._deviceId,
+    return Promise.all(promises).then(() => {
+        return crypto._baseApis.uploadKeysRequest({
```

Comment:
Given our ability to use async/await now, I would prefer it if we converted this. E.g
```js
await Promise.all(promises);
const res = await crypto._baseApis.uploadKeysRequest({
    one_time_keys: oneTimeJson,
}, {
    // for now, we set the device id explicitly, as we may not be using the
    // same one as used in login.
    device_id: crypto._deviceId,
});
crypto._olmDevice.markKeysAsPublished();
return res;
```